src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -528,6 +528,55 @@ public SequencePosition GetPosition(long offset)
             return Seek(offset);
         }
 
+        /// <summary>
+        /// Translate opaque <see cref="SequencePosition"/> value to numerical position offset
+        /// </summary>
+        /// <param name="position"></param>
+        /// <returns></returns>
+        public long GetOffset(SequencePosition position)
+        {
+            bool positionIsNotNull = position.GetObject() != null;
+            BoundsCheck(position, positionIsNotNull);
+
+            object? startObject = _startObject;
+            object? endObject = _endObject;
+
+            uint positionIndex = (uint)GetIndex(position);
+            object? positionSequenceObject = position.GetObject();
+
+            // if sequence object is null we suppose start segment
+            if (!positionIsNotNull)
```

Comment:
I used same semantic of `Slice` I don't know if it's ok, let me know

Reply:
Is there a benefit of the way `GetOffset` is implemented now instead of just doing the following:
```C#
GetOffset(SequencePosition position) => Slice(0, position).Length;
```

Reply:
@ahsonkhan did some perf test and are similar(I need to understand from where that micro allocation comes)
No Slice

|                 Method |      Mean |    Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|------:|------:|------:|----------:|
| GetOffsetSingleSegment |  73.78 ms | 0.259 ms | 0.230 ms |     - |     - |     - |     138 B |
|  GetOffsetMultiSegment | 208.02 ms | 0.669 ms | 0.626 ms |     - |     - |     - |     413 B |

Slice  


|                 Method |      Mean |    Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|------:|------:|------:|----------:|
| GetOffsetSingleSegment |  87.31 ms | 0.525 ms | 0.491 ms |     - |     - |     - |         - |
|  GetOffsetMultiSegment | 211.03 ms | 1.191 ms | 1.056 ms |     - |     - |     - |         - |

```cs
[MemoryDiagnoser]
    public class ReadOnlySequence
    {
        ReadOnlySequence<int> bufferSingleSegment;
        SequencePosition[] positionsSingleSegment;

        ReadOnlySequence<int> bufferMultiSegment;
        SequencePosition[] positionsMultiSegment;



        [GlobalSetup]
        public void Setup()
        {
            int n = 10_000_000;
            bufferSingleSegment = new ReadOnlySequence<int>(new int[n]);
            List<SequencePosition> positionsList = new List<SequencePosition>();
            for (int i = 0; i <= bufferSingleSegment.Length; i++)
            {
                positionsList.Add(bufferSingleSegment.GetPosition(i));
            }
            positionsSingleSegment = positionsList.ToArray();

            positionsList.Clear();

            var bufferSegment1 = new BufferSegment<int>(new int[n / 4]);
            BufferSegment<int> bufferSegment2 = bufferSegment1.Append(new int[n / 4]);
            BufferSegment<int> bufferSegment3 = bufferSegment2.Append(new int[n / 4]);
            BufferSegment<int> bufferSegment4 = bufferSegment3.Append(new int[n / 4]);
            bufferMultiSegment = new ReadOnlySequence<int>(bufferSegment1, 0, bufferSegment4, n / 4);
            for (int i = 0; i <= bufferMultiSegment.Length; i++)
            {
                positionsList.Add(bufferMultiSegment.GetPosition(i));
            }
            positionsMultiSegment = positionsList.ToArray();
            positionsList.Clear();
        }

        [Benchmark]
        public void GetOffsetSingleSegment()
        {
            for (int i = 0; i <= bufferSingleSegment.Length; i++)
            {
                bufferSingleSegment.GetOffset(positionsSingleSegment[i]);
            }
        }

        [Benchmark]
        public void GetOffsetMultiSegment()
        {
            for (int i = 0; i <= bufferMultiSegment.Length; i++)
            {
                bufferMultiSegment.GetOffset(positionsMultiSegment[i]);
            }
        }
    }
```
BTW the most important think is that slice doesn't work as expected on boundary and test doesn't pass for
GetOffset_MultiSegment_PositionOutOfRange_SegmentNotFound
GetOffset_MultiSegment_BoundaryConditions
GetOffset_MultiSegment_SequencePositionSegment
GetOffset_MultiSegment_PositionOutOfRange_SegmentNotFound
GetOffset_MultiSegment_BoundaryConditions
GetOffset_MultiSegment_SequencePositionSegment

What do you think?

Reply:
Interesting. I am surprised by the "no slice" method showing allocations as well. Could you measure just a single call to GetOffset rather than doing it in the loop (and let BDN deal with running it many times)?

I will take a closer look tomorrow at the tests failing if we chose to go with the simpler `Slice` approach. Maybe we make `Slice` more resilient to those edge cases.

Reply:
Letting BDN to profile e chose right iteration count

```cs
        [Benchmark]
        public void GetOffsetSingleSegment()
        {
            bufferSingleSegment.GetOffset(positionsSingleSegment[10_000]);
        }

        [Benchmark]
        public void GetOffsetMultiSegment()
        {
            bufferMultiSegment.GetOffset(positionsMultiSegment[10_000]);
        }
```

No slice
|                 Method |      Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|----------:|----------:|------:|------:|------:|----------:|
| GetOffsetSingleSegment |  6.030 ns | 0.0420 ns | 0.0393 ns |     - |     - |     - |         - |
|  GetOffsetMultiSegment | 16.933 ns | 0.0511 ns | 0.0453 ns |     - |     - |     - |         - |

Slice

|                 Method |     Mean |    Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|---------:|---------:|------:|------:|------:|----------:|
| GetOffsetSingleSegment | 15.30 ns | 0.074 ns | 0.069 ns |     - |     - |     - |         - |
|  GetOffsetMultiSegment | 22.28 ns | 0.429 ns | 0.401 ns |     - |     - |     - |         - |